README.md

# Bloaty McBloatface: a size profiler for binaries

[![Build Status](https://travis-ci.org/google/bloaty.svg?branch=master)](https://travis-ci.org/google/bloaty)

Ever wondered what's making your binary big?  Bloaty
McBloatface will show you a size profile of the binary so
you can understand what's taking up space inside.

Bloaty performs a deep analysis of the binary. Using custom
ELF, DWARF, and Mach-O parsers, Bloaty aims to accurately
attribute every byte of the binary to the symbol or
compileunit that produced it. It will even disassemble the
binary looking for references to anonymous data. For more
information about the analysis performed by Bloaty, please
see [doc/how-bloaty-works.md](doc/how-bloaty-works.md).

Bloaty works on binaries, shared objects, object files, and
static libraries (`.a` files).  The following file formats
are supported:

* ELF
* Mach-O
* WebAssembly (experimental)

These formats are NOT supported, but I am very interested
in adding support for them (I may implement these myself but
would also be happy to get contributions!)

* PE/COFF (not supported)
* Android APK (not supported, might be tricky due to compression)

This is not an official Google product.

## Building Bloaty

Building Bloaty requires CMake and ``protoc``, the protobuf compiler. On Ubuntu, install them with:

```
$ sudo apt install cmake protobuf-compiler
```

Bloaty bundles ``libprotobuf``, ``re2``, ``capstone``, and ``pkg-config`` as Git submodules, but it will prefer the system's versions of those dependencies if available. All other dependencies are included as Git submodules. To build, run:

```
$ cmake .
$ make -j6
```

To run tests (Git only, these are not included in the release tarball), type:

```
$ make test
```

All the normal CMake features are available, like out-of-source builds:

```
$ mkdir build
$ cd build
$ cmake ..
$ make -j6
```

## Running Bloaty

Run it directly on a binary target.  For example, run it on itself.

```
$ ./bloaty bloaty
```

On Linux you'll see output something like:

```cmdoutput
$ ./bloaty bloaty
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  30.0%  8.85Mi   0.0%       0    .debug_info
  24.7%  7.29Mi   0.0%       0    .debug_loc
  12.8%  3.79Mi   0.0%       0    .debug_str
   9.7%  2.86Mi  42.8%  2.86Mi    .rodata
   6.9%  2.03Mi  30.3%  2.03Mi    .text
   6.3%  1.85Mi   0.0%       0    .debug_line
   4.0%  1.19Mi   0.0%       0    .debug_ranges
   0.0%       0  15.0%  1.01Mi    .bss
   1.6%   473Ki   0.0%       0    .strtab
   1.4%   435Ki   6.3%   435Ki    .data
   0.8%   254Ki   3.7%   254Ki    .eh_frame
   0.8%   231Ki   0.0%       0    .symtab
   0.5%   142Ki   0.0%       0    .debug_abbrev
   0.2%  56.8Ki   0.8%  56.8Ki    .gcc_except_table
   0.1%  41.4Ki   0.6%  41.4Ki    .eh_frame_hdr
   0.0%  11.4Ki   0.1%  9.45Ki    [26 Others]
   0.0%  7.20Ki   0.1%  7.14Ki    .dynstr
   0.0%  6.09Ki   0.1%  6.02Ki    .dynsym
   0.0%  4.89Ki   0.1%  4.83Ki    .rela.plt
   0.0%  4.59Ki   0.0%       0    [Unmapped]
   0.0%  3.30Ki   0.0%  3.23Ki    .plt
 100.0%  29.5Mi 100.0%  6.69Mi    TOTAL
```

The "VM SIZE" column tells you how much space the binary
will take when it is loaded into memory.  The "FILE SIZE"
column tells you about how much space the binary is taking
on disk.  These two can be very different from each other:

- Some data lives in the file but isn't loaded into memory,
  like debug information.
- Some data is mapped into memory but doesn't exist in the
  file.  This mainly applies to the `.bss` section
  (zero-initialized data).

The default breakdown in Bloaty is by sections, but many
other ways of slicing the binary are supported such as
symbols and segments.  If you compiled with debug info, you
can even break down by compile units and inlines!

```cmdoutput
$ ./bloaty bloaty -d compileunits
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  34.8%  10.2Mi  43.4%  2.91Mi    [163 Others]
  17.2%  5.08Mi   4.3%   295Ki    third_party/protobuf/src/google/protobuf/descriptor.cc
   7.3%  2.14Mi   2.6%   179Ki    third_party/protobuf/src/google/protobuf/descriptor.pb.cc
   4.6%  1.36Mi   1.1%  78.4Ki    third_party/protobuf/src/google/protobuf/text_format.cc
   3.7%  1.10Mi   4.5%   311Ki    third_party/capstone/arch/ARM/ARMDisassembler.c
   1.3%   399Ki  15.9%  1.07Mi    third_party/capstone/arch/M68K/M68KDisassembler.c
   3.2%   980Ki   1.1%  75.3Ki    third_party/protobuf/src/google/protobuf/generated_message_reflection.cc
   3.2%   965Ki   0.6%  40.7Ki    third_party/protobuf/src/google/protobuf/descriptor_database.cc
   2.8%   854Ki  12.0%   819Ki    third_party/capstone/arch/X86/X86Mapping.c
   2.8%   846Ki   1.0%  66.4Ki    third_party/protobuf/src/google/protobuf/extension_set.cc
   2.7%   800Ki   0.6%  41.2Ki    third_party/protobuf/src/google/protobuf/generated_message_util.cc
   2.3%   709Ki   0.7%  50.7Ki    third_party/protobuf/src/google/protobuf/wire_format.cc
   2.1%   637Ki   1.7%   117Ki    third_party/demumble/third_party/libcxxabi/cxa_demangle.cpp
   1.8%   549Ki   1.7%   114Ki    src/bloaty.cc
   1.7%   503Ki   0.7%  48.1Ki    third_party/protobuf/src/google/protobuf/repeated_field.cc
   1.6%   469Ki   6.2%   427Ki    third_party/capstone/arch/X86/X86DisassemblerDecoder.c
   1.4%   434Ki   0.2%  15.9Ki    third_party/protobuf/src/google/protobuf/message.cc
   1.4%   422Ki   0.3%  23.4Ki    third_party/re2/re2/dfa.cc
   1.3%   407Ki   0.4%  24.9Ki    third_party/re2/re2/regexp.cc
   1.3%   407Ki   0.4%  29.9Ki    third_party/protobuf/src/google/protobuf/map_field.cc
   1.3%   397Ki   0.4%  24.8Ki    third_party/re2/re2/re2.cc
 100.0%  29.5Mi 100.0%  6.69Mi    TOTAL
```


Run Bloaty with `--help` to see a list of available options:

```cmdoutput
$ ./bloaty --help
Bloaty McBloatface: a size profiler for binaries.

USAGE: bloaty [OPTION]... FILE... [-- BASE_FILE...]

Options:

  --csv              Output in CSV format instead of human-readable.
  --tsv              Output in TSV format instead of human-readable.
  -c FILE            Load configuration from <file>.
  -d SOURCE,SOURCE   Comma-separated list of sources to scan.
  --debug-file=FILE  Use this file for debug symbols and/or symbol table.
  -C MODE            How to demangle symbols.  Possible values are:
  --demangle=MODE      --demangle=none   no demangling, print raw symbols
                       --demangle=short  demangle, but omit arg/return types
                       --demangle=full   print full demangled type
                     The default is --demangle=short.
  --disassemble=FUNCTION
                     Disassemble this function (EXPERIMENTAL)
  --domain=DOMAIN    Which domains to show.  Possible values are:
                       --domain=vm
                       --domain=file
                       --domain=both (the default)
  -n NUM             How many rows to show per level before collapsing
                     other keys into '[Other]'.  Set to '0' for unlimited.
                     Defaults to 20.
  -s SORTBY          Whether to sort by VM or File size.  Possible values
                     are:
                       -s vm
                       -s file
                       -s both (the default: sorts by max(vm, file)).
  -w                 Wide output; don't truncate long labels.
  --help             Display this message and exit.
  --list-sources     Show a list of available sources and exit.
  --source-filter=PATTERN
                     Only show keys with names matching this pattern.

Options for debugging Bloaty:

  --debug-vmaddr=ADDR
  --debug-fileoff=OFF
                     Print extended debugging information for the given
                     VM address and/or file offset.
  -v                 Verbose output.  Dumps warnings encountered during
                     processing and full VM/file maps at the end.
                     Add more v's (-vv, -vvv) for even more.

```

# Size Diffs

You can use Bloaty to see how the size of a binary changed.
On the command-line, pass `--` followed by the files you
want to use as the diff base.

For example, here is a size diff between a couple different versions
of Bloaty, showing how it grew when I added some features.

```
$ ./bloaty bloaty -- oldbloaty
     VM SIZE                     FILE SIZE
 --------------               --------------
  [ = ]       0 .debug_loc     +688Ki  +9.9%
   +19%  +349Ki .text          +349Ki   +19%
  [ = ]       0 .debug_ranges  +180Ki   +11%
  [ = ]       0 .debug_info    +120Ki  +0.9%
   +23% +73.5Ki .rela.dyn     +73.5Ki   +23%
  +3.5% +57.1Ki .rodata       +57.1Ki  +3.5%
 +28e3% +53.9Ki .data         +53.9Ki +28e3%
  [ = ]       0 .debug_line   +40.2Ki  +4.8%
  +2.3% +5.35Ki .eh_frame     +5.35Ki  +2.3%
  -6.0%      -5 [Unmapped]    +2.65Ki  +215%
  +0.5% +1.70Ki .dynstr       +1.70Ki  +0.5%
  [ = ]       0 .symtab       +1.59Ki  +0.9%
  [ = ]       0 .debug_abbrev +1.29Ki  +0.5%
  [ = ]       0 .strtab       +1.26Ki  +0.3%
   +16%    +992 .bss                0  [ = ]
  +0.2%    +642 [13 Others]      +849  +0.2%
  +0.6%    +792 .dynsym          +792  +0.6%
   +16%    +696 .rela.plt        +696   +16%
   +16%    +464 .plt             +464   +16%
  +0.8%    +312 .eh_frame_hdr    +312  +0.8%
  [ = ]       0 .debug_str    -19.6Ki  -0.4%
   +11%  +544Ki TOTAL         +1.52Mi  +4.6%
```

Each line shows the how much each part changed compared to
its previous size.  Most sections grew, but one section at
the bottom (`.debug_str`) shrank.  The "TOTAL" line shows
how much the size changed overall.

# Hierarchical Profiles

Bloaty supports breaking the binary down in lots of
different ways.  You can combine multiple data sources into
a single hierarchical profile.  For example, we can use the
`segments` and `sections` data sources in a single report:

```cmdoutput
$ ./bloaty -d segments,sections bloaty
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  80.7%  23.8Mi   0.0%       0    [Unmapped]
    37.2%  8.85Mi   NAN%       0    .debug_info
    30.6%  7.29Mi   NAN%       0    .debug_loc
    15.9%  3.79Mi   NAN%       0    .debug_str
     7.8%  1.85Mi   NAN%       0    .debug_line
     5.0%  1.19Mi   NAN%       0    .debug_ranges
     1.9%   473Ki   NAN%       0    .strtab
     1.0%   231Ki   NAN%       0    .symtab
     0.6%   142Ki   NAN%       0    .debug_abbrev
     0.0%  4.59Ki   NAN%       0    [Unmapped]
     0.0%     392   NAN%       0    .shstrtab
     0.0%     139   NAN%       0    .debug_macinfo
     0.0%      68   NAN%       0    .comment
  10.9%  3.21Mi  47.9%  3.21Mi    LOAD #4 [R]
    89.3%  2.86Mi  89.3%  2.86Mi    .rodata
     7.7%   254Ki   7.7%   254Ki    .eh_frame
     1.7%  56.8Ki   1.7%  56.8Ki    .gcc_except_table
     1.3%  41.4Ki   1.3%  41.4Ki    .eh_frame_hdr
     0.0%       1   0.0%       1    [LOAD #4 [R]]
   6.9%  2.03Mi  30.3%  2.03Mi    LOAD #3 [RX]
    99.8%  2.03Mi  99.8%  2.03Mi    .text
     0.2%  3.23Ki   0.2%  3.23Ki    .plt
     0.0%      28   0.0%      28    [LOAD #3 [RX]]
     0.0%      23   0.0%      23    .init
     0.0%       9   0.0%       9    .fini
   1.5%   439Ki  21.4%  1.44Mi    LOAD #5 [RW]
     0.0%       0  70.1%  1.01Mi    .bss
    99.1%   435Ki  29.6%   435Ki    .data
     0.4%  1.63Ki   0.1%  1.63Ki    .got.plt
     0.3%  1.46Ki   0.1%  1.46Ki    .data.rel.ro
     0.1%     560   0.0%     560    .dynamic
     0.1%     384   0.0%     376    .init_array
     0.0%      32   0.0%      56    [LOAD #5 [RW]]
     0.0%      32   0.0%      32    .got
     0.0%      16   0.0%      16    .tdata
     0.0%       8   0.0%       8    .fini_array
     0.0%       0   0.0%       8    .tbss
   0.1%  23.3Ki   0.3%  23.3Ki    LOAD #2 [R]
    30.7%  7.14Ki  30.7%  7.14Ki    .dynstr
    25.9%  6.02Ki  25.9%  6.02Ki    .dynsym
    20.8%  4.83Ki  20.8%  4.83Ki    .rela.plt
     7.7%  1.78Ki   7.7%  1.78Ki    .hash
     5.0%  1.17Ki   5.0%  1.17Ki    .rela.dyn
     3.1%     741   3.1%     741    [LOAD #2 [R]]
     2.7%     632   2.7%     632    .gnu.hash
     2.2%     514   2.2%     514    .gnu.version
     1.6%     384   1.6%     384    .gnu.version_r
     0.2%      36   0.2%      36    .note.gnu.build-id
     0.1%      32   0.1%      32    .note.ABI-tag
     0.1%      28   0.1%      28    .interp
   0.0%  2.56Ki   0.0%       0    [ELF Headers]
    46.3%  1.19Ki   NAN%       0    [19 Others]
     7.3%     192   NAN%       0    [ELF Headers]
     2.4%      64   NAN%       0    .comment
     2.4%      64   NAN%       0    .data
     2.4%      64   NAN%       0    .data.rel.ro
     2.4%      64   NAN%       0    .debug_abbrev
     2.4%      64   NAN%       0    .debug_info
     2.4%      64   NAN%       0    .debug_line
     2.4%      64   NAN%       0    .debug_loc
     2.4%      64   NAN%       0    .debug_macinfo
     2.4%      64   NAN%       0    .debug_ranges
     2.4%      64   NAN%       0    .debug_str
     2.4%      64   NAN%       0    .dynamic
     2.4%      64   NAN%       0    .dynstr
     2.4%      64   NAN%       0    .dynsym
     2.4%      64   NAN%       0    .eh_frame
     2.4%      64   NAN%       0    .eh_frame_hdr
     2.4%      64   NAN%       0    .fini
     2.4%      64   NAN%       0    .fini_array
     2.4%      64   NAN%       0    .gcc_except_table
     2.4%      64   NAN%       0    .gnu.hash
 100.0%  29.5Mi 100.0%  6.69Mi    TOTAL
```

Bloaty displays a maximum of 20 lines for each level; other
values are grouped into an `[Other]` bin.  Use `-n <num>`
to override this setting.  If you pass `-n 0`, all data
will be output without collapsing anything into `[Other]`.

# Debugging Stripped Binaries

Bloaty supports reading debuginfo/symbols from separate
binaries.  This lets you profile a stripped binary, even for
data sources like "compileunits" or "symbols" that require
this extra information.

Bloaty uses build IDs to verify that the binary and the
debug file match.  Otherwise the results would be nonsense
(this kind of mismatch might sound unlikely but it's a very
easy mistake to make, and one that I made several times even
as Bloaty's author!).

If your binary has a build ID, then using separate debug
files is as simple as:

```
$ cp bloaty bloaty.stripped
$ strip bloaty.stripped
$ ./bloaty -d symbols --debug-file=bloaty bloaty.stripped
```

Some format-specific notes follow.

## ELF

For ELF, make sure you are compiling with build IDs enabled.
With gcc this happens automatically, but [Clang decided not
to make this the default, since it makes the link
slower](http://releases.llvm.org/3.9.0/tools/clang/docs/ReleaseNotes.html#major-new-features).
For Clang add `-Wl,--build-id` to your link line.  (If you
want a slightly faster link and don't care about
reproducibility, you can use `-Wl,--build-id=uuid` instead).

Bloaty does not currently support the GNU debuglink or
looking up debug files by build ID, [which are the methods
GDB uses to find debug
files](https://sourceware.org/gdb/onlinedocs/gdb/Separate-Debug-Files.html).
If there are use cases where Bloaty's `--debug-file` option
won't work, we can reconsider implementing these.

## Mach-O

Mach-O files always have build IDs (as far as I can tell),
so no special configuration is needed to make sure you get
them.

Mach-O puts debug information in separate files which you
can create with `dsymutil`:

```
$ dsymutil bloaty
$ strip bloaty  (optional)
$ ./bloaty -d symbols --debug-file=bloaty.dSYM/Contents/Resources/DWARF/bloaty bloaty
```

# Configuration Files

Any options that you can specify on the command-line, you
can put into a configuration file instead.  Then use can use
`-c FILE` to load those options from the config file.  Also,
a few features are only available with configuration files
and cannot be specify on the command-line.

The configuration file is a in Protocol Buffers text format.
The schema is the `Options` message in
[src/bloaty.proto](src/bloaty.proto).

The two most useful cases for configuration files are:

1. You have too many input files to put on the command-line.
   At Google we sometimes run Bloaty over thousands of input
   files.  This can cause the overall command-line to exceed
   OS limits.  With a config file, we can avoid this:

   ```
   filename: "path/to/long_filename_a.o"
   filename: "path/to/long_filename_b.o"
   filename: "path/to/long_filename_c.o"
   # ...repeat for thousands of files.
   ```
2. For custom data sources, it can be very useful to put
   them in a config file, for greater reusability.  For
   example, see the custom data sources defined in
   [custom_sources.bloaty](custom_sources.bloaty).
   Also read more about custom data sources below.

# Data Sources

Bloaty has many data sources built in.  These all provide
different ways of looking at the binary.  You can also
create your own data sources by applying regexes to the
built-in data sources (see "Custom Data Sources" below).

While Bloaty works on binaries, shared objects, object
files, and static libraries (`.a` files), some of the data
sources don't work on object files.  This applies especially
to data sources that read debug info.

## Segments

Segments are what the run-time loader uses to determine what
parts of the binary need to be loaded/mapped into memory.
There are usually just a few segments: one for each set of

CHANGES.md
# Change Log

## [Bloaty McBloatface v1.1](https://github.com/google/bloaty/releases/tag/v1.1) (2020-05-23)

### Added

* **Source Filtering**: Bloaty can now filter the results based on a regex
  match. See "Source filter" in [README.md](README.md) for details. (#177)
* **Show only File or VM**: It is possible to restrict the output to show only
  "VM SIZE" or "FILE SIZE" by passing `--domain=vm` or `--domain=file`. (#165)

### Fixed

* **Use system deps in CMake build**: The CMake build can now use system deps
  for protobuf, capstone, and re2 if they are present. Hopefully this will make
  Bloaty easier to package in package managers. (#155)
* **DWARF 4 Support**: Fixed one place in Bloaty that was not accepting DWARF 4.
  (#166)
* **DWARF fix**: Added support for `DW_FORM_ref_udata`. (#197)
* **Big-endian fix**: Added missing byte swapping when getting the build ID for
  a big-endian binary. (#182)
* **WASM demangling**: WASM symbols are now properly demangled. (#149)
* **New WASM sections**: Added support for new DataCount and Event sections
  in WASM. (#178)
* **Scaling fixes**: Fixed integer overflow in 32-bit builds, and other issues
  that arise when using Bloaty with many files and/or large files. (#193)
* **Improved coverage**: Bloaty now properly attributes `.eh_frame` in object
  files, and attributes ELF headers to the corresponding section. (#168)
* **CSV escaping**: Bloaty now properly escapes CSV output when a field contains
  a comma or double quote. (#174)

### Changed

* **File size column moved left**: In the default output, the file size now
  appears on the left. This means that all numbers are now on the left, which
  leads to more readable output when viewing in a proportional font or in a
  limited-width window.

  Old:
  ```
       VM SIZE                         FILE SIZE
   --------------                   --------------
     0.0%       0 .debug_info        7.97Mi  29.5%
     0.0%       0 .debug_loc         6.40Mi  23.7%
  ```

  New:
  ```
      FILE SIZE        VM SIZE    
   --------------  -------------- 
    30.0%  8.85Mi   0.0%       0    .debug_info
    24.7%  7.29Mi   0.0%       0    .debug_loc
  ```

  This shouldn't cause breakage, as anything consuming Bloaty's output
  programmatically should be using `--csv` or `--tsv`. (#165)
* **ELF Segment labels now contain index**: Previously ELF segment labels looked
  like `LOAD [RW]` with segment flags only. Now they also contain the segment
  index, eg. `LOAD #1 [RW]`, so the output can distinguish between different
  segments with the same flags. (#159)

### Security

Bloaty should not be considered fully hardened against malicious binaries.  This
is one of the major reasons why Bloaty is not offered as an in-process library,
and should only be used through its command-line interface in a dedicated
address space. If you do not trust the input, further process sandboxing is
advisable.

However we do perform fuzzing of the parsers, and fix and crash bugs that are
found by fuzzing.

* **Fixed crash bugs found by fuzzing** (#173, #175)

## [Bloaty McBloatface v1.0](https://github.com/google/bloaty/releases/tag/v1.0) (2018-08-07)

This is the first formal release of Bloaty.

Changes since Bloaty was [first announced in Nov
2016](http://blog.reverberate.org/2016/11/07/introducing-bloaty-mcbloatface.html):

* **Much better coverage / data quality**: Bloaty now properly attributes
  sections of the binary like the symbol table, debugging information,
  relocations, and frame unwinding info. We even disassemble the binary looking
  for references to anonymous data members that do not have symbol table
  entries. This all means higher quality output, and much less of the binary is
  attributed to `[None]`.
* **Native Mach-O support**: Bloaty can now parse Mach-O files directly,
  instead of shelling out to other programs. The result is much faster and
  higher-quality output for Mach-O. Also the data sources that require debug
  info (like `-d compileunits`) now work with Mach-O.
* **WebAssembly support (EXPERIMENTAL)**: Bloaty can analyze sections and
  symbols in binary WebAssembly files.
* **Native C++ Demangling**: Bloaty can now demangle C++ symbols without
  shelling out to `c++filt`. There are two separate demangling modes, one which
  strips all template parameters and parameter/return types (`shortsymbols`) and
  one that returns full demangled names (`fullsymbols`).
* **Profiling stripped binaries**: Bloaty can read symbols and debug info from
  separate files. This allows you to profile stripped binaries.
* **Parallel file parsing**: If you pass multiple files to Bloaty, it will
  scan them in parallel. If you are parsing lots of files and have lots of CPUs,
  this can save a lot of time.
* **Configuration files**: All options you can specify on the command-line can
  be put in a configuration file instead (and vice versa). This is helpful if
  the options might otherwise overflow the command-line (for example, if you
  are parsing thousands of files). It also lets you save bits of configuration
  to a file for reuse.
* **Custom data sources**: you can create your own data sources by applying
  regexes to the built-in sources. This lets you bucket symbols, source files,
  etc. in ways that make sense for your project.
* **CSV/TSV output**: this is a robust way to parse Bloaty's output and use it
  in other programs. (The default, human-readable output is not designed to be
  parsed and its format may change in backward-incompatible ways).
* **Lots of bugfixes**: Fixed lots of bugs that people reported in various
  platforms and configurations. Bloaty is fuzzed regularly now, and this has
  led to many bugfixes also.


macho.cc
// Copyright 2016 Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <iostream>
#include "bloaty.h"
#include "re2/re2.h"

// There are several programs that offer useful information about
// binaries:
//
// - otool: display object file headers and contents (including disassembly)
// - nm: display symbols
// - size: display binary size
// - symbols: display symbols, drawing on more sources
// - pagestuff: display page-oriented information
// - dsymutil: create .dSYM bundle with DWARF information inside
// - dwarfdump: dump DWARF debugging information from .o or .dSYM

namespace bloaty {

bool StartsWith(const std::string& haystack, const std::string& needle) {
  return !haystack.compare(0, needle.length(), needle);
}

static void ParseMachOSymbols(const std::string& filename, VMRangeSink* sink) {
  std::string cmd = std::string("symbols -noSources -noDemangling ") + filename;

  // [16 spaces]0x00000001000009e0 (  0x3297) run_tests [FUNC, EXT, LENGTH, NameNList, MangledNameNList, Merged, NList, Dwarf, FunctionStarts]
  // [16 spaces]0x00000001000015a0 (     0x9) __ZN10LineReader5beginEv [FUNC, EXT, LENGTH, NameNList, MangledNameNList, Merged, NList, Dwarf, FunctionStarts]
  // [16 spaces]0x0000000100038468 (     0x8) __ZN3re2L12empty_stringE [NameNList, MangledNameNList, NList]
  RE2 pattern(R"(^\s{16}0x([0-9a-f]+) \(\s*0x([0-9a-f]+)\) (.+) \[((?:FUNC)?))");

  for ( auto& line : ReadLinesFromPipe(cmd) ) {
    std::string name;
    std::string maybe_func;
    size_t addr, size;

    if (RE2::PartialMatch(line, pattern, RE2::Hex(&addr), RE2::Hex(&size),
                          &name, &maybe_func)) {
      if (StartsWith(name, "DYLD-STUB")) {
        continue;
      }

      sink->AddVMRange(addr, size, name);
    }
  }
}

#if 0

  // file info from symbols
  // [20 spaces]0x00000001000009e0 (    0x21) tests/test_def.c:471
  RE2 pattern2(R"(^\s{20}0x([0-9a-f]+) \(\s*0x([0-9a-f]+)\) (.+):(\d+))");

static void ParseMachODisassembly(const std::string& filename,
                                  ProgramDataSink* program) {
  std::string cmd = std::string("otool -tV ") + filename;

  // ReadLinesFromPipe(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&):
  RE2 func_pattern(R"(^([a-zA-Z_].*):)");

  // 0000000100028049        callq   re2::Regexp::Incref() ## re2::Regexp::Incref()
  // Tail call:
  // 0000000100028064        jmp     re2::Regexp::Plus(re2::Regexp*, re2::Regexp::ParseFlags) ## re2::Regexp::Plus(re2::Regexp*, re2::Regexp::ParseFlags)
  RE2 call_pattern(R"((?:callq|jmp)\s+([a-zA-Z_][^#\n]*)(?: #)?)");

  // 000000010000876e        leaq    0x31e9b(%rip), %rax
  // 0000000100008775        movq    %rax, (%rbx)
  RE2 leaq_pattern(R"(leaq\s+(-?0x[0-9a-f]+)\(%rip)");
  RE2 leaq_sym_pattern(R"(leaq\s+(.+)\(%rip\))");
  RE2 addr_pattern(R"(^([0-9a-f]+))");

  Object* current_function = NULL;
  std::string current_function_name;
  long lea_offset = 0;
  std::string name;

  for ( auto& line : ReadLinesFromPipe(cmd) ) {
    size_t addr;
    if (lea_offset) {
      if (RE2::PartialMatch(line, addr_pattern, RE2::Hex(&addr))) {
        Object* dest = program->FindObjectByAddr((long)addr + lea_offset);
        if (current_function && dest) {
          program->AddRef(current_function, dest);
        }
        lea_offset = 0;
      }
    }

    if (RE2::PartialMatch(line, leaq_pattern, RE2::Hex(&lea_offset))) {
      // Next line will find the actual address.
      /*
      if (current_function) {
        std::cout << "Found lea in function; " << current_function->name << "\n";
      } else {
        std::cout << "Found lea but no current function!  Name: " << current_function_name;
      }
      */
    } else if (RE2::PartialMatch(line, func_pattern, &current_function_name)) {
      current_function = program->FindObjectByName(current_function_name);

      if (!current_function && current_function_name[0] == '_') {
        current_function_name = current_function_name.substr(1);
      }

      current_function = program->FindObjectByName(current_function_name);

      if (!current_function) {
        std::cerr << "Couldn't find function for name: " << current_function_name << "\n";
      }
    } else if (RE2::PartialMatch(line, call_pattern, &name) ||
               RE2::PartialMatch(line, leaq_sym_pattern, &name)) {
      while (name[name.size() - 1] == ' ') {
        name.resize(name.size() - 1);
      }
      Object* target_function = program->FindObjectByName(name);
      if (!target_function && name[0] == '_') {
        name = name.substr(1);
        target_function = program->FindObjectByName(name);
      }

      if (!current_function) {
        std::cerr << line;
        std::cerr << "Whoops!  Found an edge but no current function.  Name: " << name << "\n";
      } else if (!target_function) {
        std::cerr << line;
        std::cerr << "Whoops!  Couldn't find a function entry for: '" << name << "'\n";
      }

      current_function->refs.insert(target_function);

      if (current_function->file != NULL && target_function->file != NULL) {
        current_function->file->refs.insert(target_function->file);
      }
    }
  }
}
#endif

static void ParseMachOSegments(const std::string& filename,
                               VMFileRangeSink* sink) {
  // Load command 2
  //       cmd LC_SEGMENT_64
  //   cmdsize 632
  //   segname __DATA
  //    vmaddr 0x000000010003a000
  //    vmsize 0x0000000000004000
  //   fileoff 237568
  //  filesize 16384
  //   maxprot 0x00000007
  //  initprot 0x00000003
  //    nsects 7
  //     flags 0x0
  //
  std::string cmd = std::string("otool -l ") + filename;

  RE2 key_decimal_pattern(R"((\w+) ([1-9][0-9]*))");
  RE2 key_hex_pattern(R"((\w+) 0x([0-9a-f]+))");
  RE2 key_text_pattern(R"( (\w+) (\w+))");
  std::string key;
  std::string text_val;
  std::string segname;
  uintptr_t val;
  uintptr_t vmaddr = 0;
  uintptr_t vmsize = 0;
  uintptr_t fileoff = 0;
  uintptr_t filesize = 0;

  // The entry point appears to be relative to the vmaddr of the first __TEXT
  // segment.  Should verify that this is the precise rule.
  uintptr_t first_text_vmaddr = 0;
  bool in_text_section = false;

  for ( auto& line : ReadLinesFromPipe(cmd) ) {

    if (RE2::PartialMatch(line, key_decimal_pattern, &key, &val) ||
        RE2::PartialMatch(line, key_hex_pattern, &key, RE2::Hex(&val))) {
      if (key == "vmaddr") {
        vmaddr = val;
        if (first_text_vmaddr == 0 && in_text_section) {
          first_text_vmaddr = vmaddr;
        }
      } else if (key == "vmsize") {
        vmsize = val;
      } else if (key == "fileoff") {
        fileoff = val;
      } else if (key == "filesize") {
        filesize = val;
        sink->AddRange(segname, vmaddr, vmsize, fileoff, filesize);
      } else if (key == "entryoff") {
        /*
        Object* entry = sink->FindObjectByAddr(first_text_vmaddr + val);
        if (entry) {
          sink->SetEntryPoint(entry);
        }
        */
      }
    } else if (RE2::PartialMatch(line, key_text_pattern, &key, &text_val)) {
      if (key == "segname") {
        in_text_section = (text_val == "__TEXT");
        segname = text_val;
        vmaddr = 0;
        vmsize = 0;
        fileoff = 0;
        filesize = 0;
      }
    }
  }
}

static void ParseMachOSections(const std::string& filename,
                               VMFileRangeSink* sink) {
  // Section
  //   sectname __text
  //    segname __TEXT
  //       addr 0x0000000100000ac0
  //       size 0x0000000000030b10
  //     offset 2752
  //      align 2^4 (16)
  //     reloff 0
  //     nreloc 0
  //      flags 0x80000400
  //  reserved1 0
  //  reserved2 0

  std::string cmd = std::string("otool -l ") + filename;

  RE2 key_decimal_pattern(R"((\w+) ([1-9][0-9]*))");
  RE2 key_hex_pattern(R"((\w+) 0x([0-9a-f]+))");
  RE2 key_text_pattern(R"( (\w+) (\w+))");
  std::string key;
  std::string text_val;
  std::string sectname;
  std::string segname;
  uintptr_t val;
  uintptr_t addr = 0;
  uintptr_t size = 0;
  uintptr_t offset = 0;

  for ( auto& line : ReadLinesFromPipe(cmd) ) {

    if (RE2::PartialMatch(line, key_decimal_pattern, &key, &val) ||
        RE2::PartialMatch(line, key_hex_pattern, &key, RE2::Hex(&val))) {
      if (key == "addr") {
        addr = val;
      } else if (key == "size") {
        size = val;
      } else if (key == "offset") {
        offset = val;
      } else if (key == "size") {
        size = val;
      } else if (key == "flags") {
        size_t filesize = size;
        if (val & 0x1) {
          filesize = 0;
        }

        if (segname.empty() || sectname.empty()) {
          continue;
        }

        std::string label = segname + "," + sectname;
        sink->AddRange(label, addr, size, offset, filesize);

        sectname.clear();
        segname.clear();
        addr = 0;
        size = 0;
        offset = 0;
      }
    } else if (RE2::PartialMatch(line, key_text_pattern, &key, &text_val)) {
      if (key == "sectname") {
        sectname = text_val;
      } else if (key == "segname") {
        segname = text_val;
      }
    }
  }
}

void RegisterMachODataSources(std::vector<DataSource>* sources) {
  sources->push_back(VMFileRangeDataSource("segments", ParseMachOSegments));
  sources->push_back(VMFileRangeDataSource("sections", ParseMachOSections));
  sources->push_back(VMRangeDataSource("symbols", ParseMachOSymbols));
}

}  // namespace bloaty
